### ISSUE TITLE
no-undef doesn't understand defining while exporting

### ISSUE THREAD

**Issue by ffxsam** [eslint/eslint#5472]
Using version 2.2.0. no-undef complains about the following:

```js
export default ArtistSettings = (props) => {
  return <div>
  </div>
}
```

It complains that `ArtistSettings` is not defined, so it doesn't seem to understand that that's where I'm defining it! :)

**Comment by platinumazure** [comment]
Do you need the variable name there? Since it's a default export, couldn't you export the arrow function directly? I think that's what the rule is complaining about- you're exporting an AssignmentExpression which is assigning to a variable that was never declared.

**Comment by ffxsam** [comment]
The name is needed. Here's the full code:

```jsx
import React, {PropTypes} from 'react';

export default ArtistSettings = (props) => {
  return <div>
  </div>
}

ArtistSettings.propTypes = {};
```

The alternative solution to make ESlint happy would be:

```
import React, {PropTypes} from 'react';

const ArtistSettings = (props) => {
  return <div>
  </div>
};

ArtistSettings.propTypes = {};

export default ArtistSettings
```

**Comment by mysticatea** [comment]
I think:

```js
export default ArtistSettings = (props) => {
```
```js
export default const ArtistSettings = (props) => {
```

**Comment by ffxsam** [comment]
export default const ArtistSettings = (props) => {

That's invalid syntax.

**Comment by mysticatea** [comment]
Ah, you are right.
So you can use FunctionDeclaration instead of ArrowFunction.

**Comment by ffxsam** [comment]
Turning `=>` into `function` makes no difference either.

**Comment by mysticatea** [comment]
Sorry, I meant:

```js
export default function ArtistSettings(props) {
```

**Comment by kaicataldo** [comment]
@ffxsam Curious if @mysticatea's suggestion worked - this is how I would handle it too!

**Comment by ffxsam** [comment]
Yes, it does. But you should be able to use arrow functions, though. To complain about something not being defined, which is clearly being defined, is a bug in eslint IMO.

**Comment by kaicataldo** [comment]
Hm, maybe I'm missing something, but I agree with @platinumazure's comment above:
> you're exporting an AssignmentExpression which is assigning to a variable that was never declared

If you disagree, mind explaining why? The problem doesn't seem to be with the arrow function. According to your comment above, this would also warn, right?
```js
export default ArtistSettings = function(props) {
  return <div>
  </div>
}
```
The problem seems to me that ArtistSettings is never declared

**Comment by ffxsam** [comment]
But if you're declaring it while exporting it, is that not declaring it?

    export default ArtistSettings = ...

**Comment by michaelficarra** [comment]
That's not a declaration.

**Comment by ffxsam** [comment]
Ok, so it’s an expression. But the eslint rule is called `no-undef`, not `no-undecl`.

Can we stop splitting hairs here? The fact is that the rule has a bug.

**Comment by michaelficarra** [comment]
@ffxsam It most certainly doesn't. An assignment is very different from a declaration. In fact, you should be getting a runtime ReferenceError because of your assignment to an undeclared variable in strict mode. In other words, your program wouldn't even work.

**Comment by ffxsam** [comment]
Wow.. ok, you're right on all fronts. My apologies. I have the following code:

```
'use strict';

import React, { PropTypes } from 'react';
import InsightDetailContainer from
  '/features/Insights/client/containers/InsightDetailContainer';

export default InsightDetailPage = (props) => {
  return <div className="container page">
    <InsightDetailContainer insightId={props.insightId} />
  </div>
}

InsightDetailPage.propTypes = {
  insightId: PropTypes.string.isRequired,
};
```

And with strict mode, it complains that `InsightDetailPage` is not being exported. Actually, I get two different errors. In my IDE, it complains it's not being exported, and in the browser console it says it's not defined.

I suppose I'm curious now why `export default` doesn't allow you to declare a variable in one shot as you can with `export`.
